File: bootstrap/src/main/java/net/dontdrinkandroot/wicket/bootstrap/component/nav/AbstractRepeatingNav.java
/*
 * Copyright (C) 2012-2017 Philip Washington Sorst <philip@sorst.net>
 * and individual contributors as indicated
 * by the @authors tag.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package net.dontdrinkandroot.wicket.bootstrap.component.nav;

import net.dontdrinkandroot.wicket.behavior.CssClassAppender;
import net.dontdrinkandroot.wicket.bootstrap.component.item.ItemContainer;
import net.dontdrinkandroot.wicket.bootstrap.css.BootstrapCssClass;
import net.dontdrinkandroot.wicket.css.CssClass;
import org.apache.wicket.markup.html.panel.GenericPanel;
import org.apache.wicket.markup.repeater.RepeatingView;
import org.apache.wicket.model.IModel;

/**
 * @author Philip Washington Sorst <philip@sorst.net>
 */
public class AbstractRepeatingNav<T> extends GenericPanel<T> implements ItemContainer
{
    public AbstractRepeatingNav(String id)
    {
        super(id);
    }

    public AbstractRepeatingNav(String id, IModel<T> model)
    {
        super(id, model);
    }

    @Override
    public CssClass getItemClass()
    {
        return BootstrapCssClass.NAV_ITEM;
    }

    @Override
    public CssClass getLinkClass()
    {
        return BootstrapCssClass.NAV_LINK;
    }

    @Override
    protected void onInitialize()
    {
        super.onInitialize();

        this.add(new CssClassAppender(BootstrapCssClass.NAV));

        RepeatingView itemView = new RepeatingView("item");
        this.populateItems(itemView);
        this.add(itemView);
    }

    protected void populateItems(RepeatingView itemView)
    {
        /* Hook */
    }
}

File: bootstrap/src/main/kotlin/net/dontdrinkandroot/wicket/bootstrap/component/nav/AbstractRepeatingNav.kt
package net.dontdrinkandroot.wicket.bootstrap.component.nav

import net.dontdrinkandroot.wicket.bootstrap.component.item.ItemContainer
import net.dontdrinkandroot.wicket.bootstrap.component.item.ItemView
import net.dontdrinkandroot.wicket.bootstrap.css.BootstrapCssClass
import net.dontdrinkandroot.wicket.css.CssClass
import org.apache.wicket.behavior.Behavior
import org.apache.wicket.markup.html.panel.GenericPanel
import org.apache.wicket.model.IModel

abstract class AbstractRepeatingNav<T>(
    id: String,
    model: IModel<T>? = null,
    vararg behaviors: Behavior
) : GenericPanel<T>(id, model), ItemContainer {

    init {
        add(*behaviors)
    }

    override val itemClass: CssClass
        get() = BootstrapCssClass.NAV_ITEM

    override val linkClass: CssClass
        get() = BootstrapCssClass.NAV_LINK

    override fun onInitialize() {
        super.onInitialize()
        val itemView = ItemView("item")
        populateItems(itemView)
        this.add(itemView)
    }

    abstract fun populateItems(itemView: ItemView)
}